From 575a1d4bdfa2ea9fc10733013136145b497e1be0 Mon Sep 17 00:00:00 2001
From: Jiaying Zhang <jiayingz@google.com>
Date: Sun, 10 Jul 2011 20:07:25 -0400
Subject: [PATCH] ext4: free allocated and pre-allocated blocks when check_eofblocks_fl fails
Patch-mainline: 575a1d4bdfa2ea9fc10733013136145b497e1be0
References: 

Upon corrupted inode or disk failures, we may fail after we already
allocate some blocks from the inode or take some blocks from the
inode's preallocation list, but before we successfully insert the
corresponding extent to the extent tree. In this case, we should free
any allocated blocks and discard the inode's preallocated blocks
because the entries in the inode's preallocation list may be in an
inconsistent state.

Signed-off-by: Jiaying Zhang <jiayingz@google.com>
Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
Cc: stable@kernel.org
Acked-by: Tao Ma <boyu.mt@taobao.com>

---
 fs/ext4/extents.c |    7 +++----
 1 files changed, 3 insertions(+), 4 deletions(-)

Index: linux-2.6.32-220.0.0.el5/fs/ext4/extents.c
===================================================================
--- linux-2.6.32-220.0.0.el5.orig/fs/ext4/extents.c	2011-12-13 14:15:21.435062221 +0800
+++ linux-2.6.32-220.0.0.el5/fs/ext4/extents.c	2011-12-13 14:29:45.639347577 +0800
@@ -3525,10 +3525,9 @@
 	}
 
 	err = check_eofblocks_fl(handle, inode, iblock, path, ar.len);
-	if (err)
-		goto out2;
-
-	err = ext4_ext_insert_extent(handle, inode, path, &newex, flags);
+	if (!err)
+		err = ext4_ext_insert_extent(handle, inode, path,
+					     &newex, flags);
 	if (err) {
 		/* free data blocks we just allocated */
 		/* not a good idea to call discard here directly,
